flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java

```diff
@@ -97,6 +97,11 @@ public void init() throws Exception {
         inputProcessor = new StreamOneInputProcessor<>(input, output, operatorChain);
     }
 
+    @Override
+    protected void finishTask() throws Exception {
+        stopMailboxProcessor();
```

Comment:
A discussion between me and @pnowojski in FLINK-21821 convinces me that `cancelables` should resides in two code path `cancel` and `cleanUpInvoke` to cancel/cleanup possible background activities. I think `cancelables.close` in `finishTask` is unnecessary at least, possibly not harmful also. I would like to not call `cancelables.close` in new code path, at least in this fix.

Reply:
I think I agree with @kezhuw , but I'm worried this is just a bug and it would brake clean shutdown of `stop-with-savepoint WITH drain`. As far as I can tell, `cancelables.close()` is closing some state backend resources. If they are closed, I wouldn't be surprised that operators/user functions would start failing during the clean shutdown. Calling `mailboxProcessor.allActionsCompleted()` here should be enough.

Probably this is missing a test coverage.

Reply:
I thought a bit about this. In normal path(same as no canceling path currently), `cancelables.close` should be called after `afterInvoke` where outputs are flushed, operators are closed/disposed. I believe stop-with-savepoint belongs to normal path(at least no canceling path). So, I tend to not call `cancelables.close` in `finishTask`.

Reply:
@kezhuw @pnowojsk Thanks for the comment. I have misunderstood the usage of cancelables. From what Piotr describes, it seems to me that the cancelables should only be closed in the common cleanup phase along with operator disposal. `cleanupInvoke` is already doing that. 

The most counter-intuitive part of the procedure here is that `StreamTask.invoke()` throws `CancelTaskException` upon cancelation without going through the cleanup logic... If the reason behind this is to allow fast shutdown in cases that a job running in a dedicated cluster is canceled, so that no cleanup is needed because JVM will exit, maybe we should have a flag like SKIP_CLEANUP_ON_CANCELATION.

For this patch, I'll just remove the `cancelable.close()` section here.

Reply:
> it would brake clean shutdown of stop-with-savepoint WITH drain

Agree. If not currently, then it will break when savepoint for drain moving after end-input-flushing operation. In that case, the whole task should function normally just no input anymore.

Reply:
>  is that StreamTask.invoke() throws CancelTaskException upon cancelation without going through the cleanup logic... 

? `cleanUpInvoke()` is/should always be called, regardless of clean shutdown or abrupt cancelation.

Reply:
@becketqin `CancelTaskException` is caught in `catch` clause later inside `StreamTask.invoke`. `cleanUpInvoke` is invoked anyway though `CancelTaskException` is propagated to `Task` for state handling.

Reply:
@pnowojski @kezhuw You are right. I misread the code. If so, why do we still need to call `cancelables.close()` in `cancel()` when the completion future is completed? What is the concern of removing that section?

Reply:
@becketqin I raised similar concern in FLINK-21821 😄 . Here is thoughts from my side after digging history a bit. Normal shutdown runs `afterInvoke`, `cleanUpInvoke` in order. `cancel` is responsible for abnormal path, which includes:

* User don't want a normal shutdown in willing.
* User has to do this due to stuck in normal shutdown.

In the later path, this `cancelables.close` and other closing operations(`TaskCanceler.networkResourcesCloser`) in cancel path plays crucial role to break possible stuck.

There might be other concerns such as fast shutdown of background activities though.

Anyway,`cancel` has to deal with un-happy paths.

Reply:
If you go through the annotations/git blame of the cancellables and how were they introduced (you need to go through two or three levels of changes), you will find this: https://issues.apache.org/jira/browse/FLINK-3466

and commit message:
```
[FLINK-3466] [runtime] Make state handles cancelable.  

State handles are cancelable, to make sure long running checkpoint restore operations do finish early on cancallation, even if the code does not properly react to interrupts.

This is especially important since HDFS client code is so buggy that it deadlocks when interrupted without closing.
```